src/MDOP.jl

```diff
@@ -87,3 +220,134 @@ function beta_statistic(Y::Dataset, s::Vector; τs = 0:50 , w::Int = 1)
+distances of the nearest neighbor for each fiducial point in one embedding
+dimension higher using a given `τ`. The obligatory threshold `r` is by default
+set to 2.
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+"""
+function fnn(NNdist::T,NNdistnew::T;r::Float64=2.0) where {T}
+    @assert length(NNdist) == length(NNdistnew) "Both input vectors need to store the same number of distances."
+
+    # convert array of arrays into simple vectors, since we only look at K=1
+    NN_old = zeros(length(NNdist))
+    NN_new = zeros(length(NNdistnew))
+    for i = 1:length(NNdist)
+        NN_old[i]=NNdist[i][1]
+        NN_new[i]=NNdistnew[i][1]
+    end
+
+    # first ratio
+    ratio1 = 1/r    # since we input only z-standardized data with unit
+    # construct statistic
+    ratio2 = NN_new./NN_old
+
+    cond1 = ratio2 .> r
+    cond2 = NN_old .< ratio1
+
+    fnns = sum(cond1.*cond2)
+    fnns2 = sum(NN_old .< ratio1)
+
+    # store fraction of valid nearest neighbors
+    if fnns2 == 0
+        FNN = NaN;
+    else
+        FNN = fnns/fnns2
+    end
+
+    return FNN
+end
+
+
+"""
+    estimate_maximum_delay(s::Vector; tw=1:50, samplesize::Float=1.0)) -> τ_max, L
+computes an upper bound for the search of optimal delays, when using `MDOP`
+[`MDOP`](@ref) or `beta_statistic` [`beta_statistic`](@ref).
+
+## Description
+The input time series `s` gets embedded with unit lag and increasing dimension,
+for dimensions (or time windows) `tw` (`RangeObject`).
+For each of such a time window the `L`-statistic from Uzal et al. [^Uzal2011]
+will be computed. `samplesize` determines the fraction of points to be
+considered in the computation of `L` (see [`uzal_cost`](@ref)). When this
+statistic reaches its global minimum the maximum delay value `τ_max` gets
+returned. When `s` is a multivariate `Dataset`, `τ_max` will becomputed for all
+time series of that Dataset and the maximum value will be returned. The returned
+`L`-statistic is a Dataset of size (length(tw)*size(s,2)).
+
+[^Nichkawde2013]: Nichkawde, Chetan (2013). [Optimal state-space reconstruction using derivatives on projected manifold. Physical Review E 87, 022905](https://doi.org/10.1103/PhysRevE.87.022905).
+[^Uzal2011]: Uzal, L. C., Grinblat, G. L., Verdes, P. F. (2011). [Optimal reconstruction of dynamical systems: A noise amplification approach. Physical Review E 84, 016223](https://doi.org/10.1103/PhysRevE.84.016223).
+"""
+function estimate_maximum_delay(s::Vector{Float64}; tw=1:50, samplesize::Float64=1.0)
+    L = zeros(Float64,length(tw),1)
+    # loop over time windows
+    cnt = 1
+    for i in tw
+        Y_act = embed(s,i,1)
+        L[cnt] = uzal_cost(Y_act; samplesize = samplesize)
+        cnt +=1
+    end
+    ~, τ_max = findmin(L)
+    return tw[τ_max], Dataset(L)
+end
```

Comment:
I now see the problem. Yes, here the creation `Dataset(L)` is indeed type unstable. Why? Well, because the **size** of a `Matrix` in Julia is not known by its type. a `Matrix{Float64}` can be 10 by 10 or it can be 2348 by 217 and still have the same type.

But more importantly, here you should not make `L` a matrix. It is just a vector, since its second dimension has size one. Why not just do `L = zeros(T, length(tw))`? Then doing `Dataset(L)` works just as well and is type stable. You can check this by doing:
```julia
@code_warntype Dataset(rand(100))
```